clang/lib/CIR/CodeGen/CIRGenTypes.cpp

```diff
@@ -18,6 +21,87 @@ mlir::MLIRContext &CIRGenTypes::getMLIRContext() const {
   return *builder.getContext();
 }
 
+/// Return true if the specified type in a function parameter or result position
+/// can be converted to a CIR type at this point. This boils down to being
+/// whether it is complete, as well as whether we've temporarily deferred
+/// expanding the type because we're in a recursive context.
+bool CIRGenTypes::isFuncParamTypeConvertible(clang::QualType type) {
+  // Some ABIs cannot have their member pointers represented in LLVM IR unless
+  // certain circumstances have been reached.
+  assert(!type->getAs<MemberPointerType>() && "NYI");
+
+  // If this isn't a tagged type, we can convert it!
```

Comment:
First, `tag` type, not `tagged` type, as this just represents any of the struct, class, union, or enum types.

Second, is the rest of this true?  Non-Tag types include atomics, Attributed types, bitint types, block pointers, complex, various dependent types, vector types, etc.  Do we really just mean to do `BuiltinType` here instead?  Or do we really mean we can do the rest already?

Reply:
This function is used to check if a type is problematic when it is used as part of a function type.  According to the comment here, only incomplete `class`/`struct`/`union` types are problematic when creating ClangIR function types.

The currently upstreamed ClangIR code can't handle atomic types or vector types or some of the other types you mentioned, in any context.  But when ClangIR can handle them, they will just work as part of function types; the code here won't need to change.

This entire function could have been left out of this PR since it doesn't do anything useful yet.  But because everything in the function except for the last line builds and runs, I decided to include it in this PR.

Reply:
In that case, I think just an assert in the  '!TagType' check here would be valuable, just that asserts it is a `BuiltinType` (or, the NYI diagnostic).

Reply:
I disagree about this one.  If the source code uses a vector type as the type of a function parameter, ClangIR code gen will fail with a NYI error in `CIRGenTypes::convertType` when trying to convert the Clang vector type into a CIR type.  There isn't a good reason to add a general-purpose check here in a function that is looking for a very specific condition that has only to do with function types.

Maybe this function should be renamed to `isConvertedTypeUsableInFunctionTypes`, to emphasize that it is not checking if a type is convertible, but checking whether or not a type is problematic if used as a return type or a parameter type within a function type.

Reply:
IMO, the assert here makes a lot of sense, in that it shows off the assumption that we're intentionally not handling these types.  The point is that we're NOT doing the work to figure out if those types are problematic if used in return /parameter in a function type (for good reason, since it'll fail earlier).  SO we should assert for that, so that when `CIRGenTypes::convertType` stops emitting the NYI, that this will be quickly noticed, rather than returning a value we wouldn't be sure of.

Reply:
>It is known that incomplete tag types are the only ones that are problematic in this context.

This seems inaccurate to me.  CIR doesn't seem to be able to represent(see here: https://github.com/llvm/clangir/blob/main/clang/include/clang/CIR/Dialect/IR/CIRTypes.td)  Atomic Types, BitInt types, Block types, and Matrix types upon quick look (that is perhaps an incomplete list).  

The function was written across; 
https://github.com/llvm/clangir/commit/101e732565d0e5004c5191b0053eebd8e849cade
and 
https://github.com/llvm/clangir/commit/93f55f6d43be394250009013c83d343be7c63f41

Neither of which mention those types specifically?  It seems it USED to have an assert for not-tag-type, though it isn't clear that the authors considered the above types.

Can @lanza and @bcardosolopes chime in and let me know if I'm misinterpreting the commits there?

Reply:
I think it's totally fine to be conservative about assumptions, we need asserts for both things that are NYI but also for things we assume as invariants, IMO the more the merrier cause it's easier to catch miscompilations.

> Neither of which mention those types specifically? It seems it USED to have an assert for not-tag-type, though it isn't clear that the authors considered the above types.

I don't remember the full context anymore, but assertion on a assumed invariant in the incubator to be even more protective also sounds good to me (guess running the tests upon such change will give us the final answer).

Reply:
@bcardosolopes What exactly do you suggest doing here in this function.  Asserting on assumed invariants is great.  But I don't see any invariants here that could be asserted.

Reply:
I don't remember the details on why it was necessary at the time to deviate - so I just made this be equivalent to traditional codegen and everything passes: https://github.com/llvm/clangir/commit/04d7dcfb2582753f3eccbf01ec900d60297cbf4b

That said, I don't think we need to introduce any asserts here, can't get better than OG for the sake of completeness. Not supported types will crash given the other mechanisms David mentioned, but the recursive logic is sound with pre-existing work.

Unless Erich has any other concerns, this LGTM to land.

Reply:
If this includes llvm/clangir@04d7dcf in it, rather than the current implementation, I think I'm OK without any asserts.  I'll Approve, but conditional on that patch being included in this one.